### PR TITLE
[Feature] support registering partial functions and more

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

**This PR is converted to DRAFT because I think this may break the design of `Registry` and may mess up some concept. I will re-examine the necessity and rationality of this change.**

Support registering partial functions in MMEngine, solve #594 

This feature might be required by FSDP for `auto_wrap_policies`

## Modification

modify `Registry` to accept all `Callable`, including classes, functions, partial functions and more

## BC-breaking (Optional)

No

## Use cases (Optional)

```python
import functools
from mmengine.registry import Registry
FUNCS = Registry('funcs')
@FUNCS.register_module('resnet')
def build_resnet(layers=50, activation='relu'):
    pass
build_resnet18 = functools.partial(build_resnet, layers=18)
FUNCS.register_module('resnet18', False, build_resnet18)
model = FUNCS.build(dict(type='resnet18', activation='silu'))
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/registry/registry.py [hunk 1]
```diff
@@ -441,16 +444,16 @@ def _register_module(self,
         """Register a module.
 
         Args:
-            module (type): Module class or function to be registered.
+            module (type): Module to be registered. Typically a class or a
+                function, but generally all `Callable` are acceptable.
```
Comment (HAOCHENYE) [inline mmengine/registry/registry.py:448]:
```suggestion
                function, but generally all ``Callable`` are acceptable.
```

File: mmengine/registry/build_functions.py [hunk 2]
```diff
@@ -120,10 +121,15 @@ def build_from_cfg(
             else:
                 obj = obj_cls(**args)  # type: ignore
 
+            # For some rare cases (e.g. obj_cls is a partial function), obj_cls
+            # doesn't have the following attributes. Use default value to
+            # prevent error
+            cls_name = getattr(obj_cls, '__name__', str(obj_cls))
+            cls_module = getattr(obj_cls, '__module__', 'unknown')
```
Comment (zhouzaida) [inline mmengine/registry/build_functions.py:128]: Can we use `isinstance(obj_cls, functools.partial)` to check whether it is a partial function? If it is a partial function, we can get its name and module from `obj_cls.func`.

Comment (C1rN09) [reply mmengine/registry/build_functions.py:128]: Before this PR, `Registry` only support functions and classes, whose definition/implementation can all be found in their `__module__`. This is good for debugging.

However, this PR allows registering generic callable `object`, whose debug info should be where/how they are created. As far as I know, this info seems only accessible for users who created this `object`.

Comment (C1rN09) [reply mmengine/registry/build_functions.py:128]: For partial functions, `str(obj_cls)` seems to contain enough debug information.
```python
>>> def foo(x):
...     pass
... 
>>> bar=functools.partial(foo, x=0)
>>> bar
functools.partial(<function foo at 0x7f8fd908a7a0>, x=0)
>>> str(bar)
'functools.partial(<function foo at 0x7f8fd908a7a0>, x=0)'
```

Comment (zhouzaida) [reply mmengine/registry/build_functions.py:128]: Got it

**Review by zhouzaida — APPROVED** [review]

**Review by mzr1996 — APPROVED** [review]

File: mmengine/registry/build_functions.py [hunk 3]
```diff
@@ -120,12 +121,18 @@ def build_from_cfg(
...
+                print_log(
+                    f'An `{obj_cls.__name__}` instance is built from '  # type: ignore # noqa: E501
+                    'registry, and its implementation can be found in '
+                    f'{obj_cls.__module__}',  # type: ignore
+                    logger='current',
+                    level=logging.DEBUG)
+            else:
+                print_log(
+                    'An instance is built from registry, and its constructor '
+                    f'is {obj_cls}')
```
Comment (zhouzaida) [inline mmengine/registry/build_functions.py:135]:
```suggestion
                    f'is {obj_cls}',
                    logger='current',
                    level=logging.DEBUG))
```

**Review by zhouzaida — APPROVED** [review]

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**